# Project: Wrangling and Analyze Data

## Data Gathering

1. Directly download the WeRateDogs Twitter archive data (twitter_archive_enhanced.csv)

In [1]:
import pandas as pd
import glob
import json
import numpy as np

In [2]:
df = pd.read_csv('twitter-archive-enhanced.csv')
df

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

2. Use the Requests library to download the tweet image prediction (image_predictions.tsv)

In [3]:
import requests
response = requests.get("https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv")

file = open("image_predictions.tsv", "wb")
file.write(response.content)
file.close() 

In [4]:
df_image = pd.read_csv('image-predictions.tsv', sep='\t')

In [5]:
df_image.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

3. Use the Tweepy library to query additional data via the Twitter API (tweet_json.txt)

In [6]:
tweets_list = []
with open('tweet-json.txt', 'r') as file:
    for line in file:
        data = json.loads(line)
        tweets_list.append(data)
df_tweet = pd.DataFrame(tweets_list, columns=['id', 'retweet_count', 'favorite_count'])

In [7]:
df_tweet.head()

id  retweet_count  favorite_count
0  892420643555336193           8853           39467
1  892177421306343426           6514           33819
2  891815181378084864           4328           25461
3  891689557279858688           8964           42908
4  891327558926688256           9774           41048

In [8]:
df_tweet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
id                2354 non-null int64
retweet_count     2354 non-null int64
favorite_count    2354 non-null int64
dtypes: int64(3)
memory usage: 55.2 KB


## Assessing Data
In this section, detect and document at least **eight (8) quality issues and two (2) tidiness issue**. You must use **both** visual assessment
programmatic assessement to assess the data.

**Note:** pay attention to the following key points when you access the data.

* You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This [unique rating system](http://knowyourmeme.com/memes/theyre-good-dogs-brent) is a big part of the popularity of WeRateDogs.
* You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.



In [9]:
df

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [11]:
df.duplicated()

0       False
1       False
2       False
3       False
4       False
5       False
6       False
7       False
8       False
9       False
10      False
11      False
12      False
13      False
14      False
15      False
16      False
17      False
18      False
19      False
20      False
21      False
22      False
23      False
24      False
25      False
26      False
27      False
28      False
29      False
        ...  
2326    False
2327    False
2328    False
2329    False
2330    False
2331    False
2332    False
2333    False
2334    False
2335    False
2336    False
2337    False
2338    False
2339    False
2340    False
2341    False
2342    False
2343    False
2344    False
2345    False
2346    False
2347    False
2348    False
2349    False
2350    False
2351    False
2352    False
2353    False
2354    False
2355    False
Length: 2356, dtype: bool

In [12]:
df.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [13]:
df.puppo.value_counts()

None     2326
puppo      30
Name: puppo, dtype: int64

In [14]:
df.name.nunique

<bound method IndexOpsMixin.nunique of 0        Phineas
1          Tilly
2         Archie
3          Darla
4       Franklin
5           None
6            Jax
7           None
8           Zoey
9         Cassie
10          Koda
11         Bruno
12          None
13           Ted
14        Stuart
15        Oliver
16           Jim
17          Zeke
18       Ralphus
19        Canela
20        Gerald
21       Jeffrey
22          such
23        Canela
24          None
25          None
26          Maya
27        Mingus
28         Derek
29        Roscoe
          ...   
2326       quite
2327           a
2328        None
2329        None
2330        None
2331        None
2332        None
2333          an
2334           a
2335          an
2336        None
2337        None
2338        None
2339        None
2340        None
2341        None
2342        None
2343        None
2344        None
2345         the
2346         the
2347           a
2348           a
2349          an
2350           a
2351     

In [15]:
df_image

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [16]:
df_image.img_num.value_counts()

1    1780
2     198
3      66
4      31
Name: img_num, dtype: int64

In [17]:
df_image.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


### Quality issues
1. there are retweets in the data

2. missing and incorrect names such as (the, a, none)

3. some values in the denominator ratings are incorrect

4. the dog stages are in object instead of boolean

5. the dog stages variables are in different column

6. ratings are in two columns

7. some images are not dogs

8. timestamp is in string format

### Tidiness issues
1. favorite count and retweet count should be part of the main dataframe

2. Timestamp contains two variables; date and time

## Cleaning Data
In this section, clean **all** of the issues you documented while assessing. 

**Note:** Make a copy of the original data before cleaning. Cleaning includes merging individual pieces of data according to the rules of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html). The result should be a high-quality and tidy master pandas DataFrame (or DataFrames, if appropriate).

In [18]:
# Make copies of original pieces of data
df_clean = df.copy()
df_tweetclean = df_tweet.copy()
df_imageclean = df_image.copy()

### Issue #1: There are retweets in the actual data

#### Define: Drop the values that are not null in the retweeted_status_id

#### Code

In [19]:
df_clean = df_clean[~df_clean['retweeted_status_id'].notnull()]
print(df_clean)

                tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                 

#### Test

In [20]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2175 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2175 non-null object
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null int64
rating_denominator            2175 non-null int64
name                          2175 non-null object
doggo                         2175 non-null object
floofer                       2175 non-null object
pupper                        2175 non-null object
puppo                         2175 non-null object
dtypes: float64(4), int64(3), object(1

### Issue #2: Wrong denominator values in some rows

#### Define: All denominator value will be replace with the standard value 10

###### Code 

In [21]:
df_clean['rating_denominator'] = df['rating_denominator'].apply(lambda x: 10)

#### Test

In [22]:
df_clean.rating_denominator.value_counts()

10    2175
Name: rating_denominator, dtype: int64

In [23]:
df_clean.head(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

### Issue #3: Missing and incorrect names

#### Define: Gather the wrong names and replace with general dog name

##### Code 

In [24]:
df_clean['name'] = df_clean['name'].replace(['the','a','an','None','quite','such'],'Dog')

#### Test

In [25]:
df_clean.name.value_counts()

Dog          753
Charlie       11
Lucy          11
Oliver        10
Cooper        10
Penny          9
Tucker         9
Lola           8
Winston        8
Sadie          8
Daisy          7
Toby           7
Stanley        6
Oscar          6
Jax            6
Bo             6
Koda           6
Bella          6
Bailey         6
Dave           5
Louis          5
Rusty          5
Chester        5
Scout          5
Bentley        5
Leo            5
Milo           5
Buddy          5
Finn           4
Gus            4
            ... 
Brian          1
Dutch          1
Billl          1
Simba          1
my             1
Shooter        1
Cupid          1
Chompsky       1
Stewie         1
Dex            1
Wafer          1
Ralf           1
Strudel        1
Ruffles        1
Jeb            1
Robin          1
Lassie         1
Gromit         1
Ronnie         1
Brady          1
Oddie          1
Brat           1
Stuart         1
Jessiga        1
Callie         1
Kathmandu      1
Bloop          1
Striker       

### Issue 4: Dog Stages are in Strings instead of boolean 

#### Define: Convert all the stages column from strings to boolean 

##### Code 

In [26]:
df_clean['doggo'] = df_clean['doggo'].map({'doggo': True, 'None': False})
df_clean['floofer'] = df_clean['floofer'].map({'floofer': True, 'None': False})
df_clean['pupper'] = df_clean['pupper'].map({'pupper': True, 'None': False})
df_clean['puppo'] = df_clean['puppo'].map({'puppo': True, 'None': False})

##### Test 

In [27]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2175 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2175 non-null object
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null int64
rating_denominator            2175 non-null int64
name                          2175 non-null object
doggo                         2175 non-null bool
floofer                       2175 non-null bool
pupper                        2175 non-null bool
puppo                         2175 non-null bool
dtypes: bool(4), float64(4), int64(3), object(

### Issue 5: The dog stages variables are in different columns  

#### Define: Aggregate the four columns into one column named Stages 

##### Code 

In [28]:
dog_stages = df_clean[['doggo','floofer','pupper','puppo']]

def dog_stages(row):
   if row['doggo'] == True:
      return 'doggo'
   if row['floofer'] == True:
      return 'floofer'
   if row['pupper'] == True:
      return 'pupper'
   if row['puppo'] == True:
      return 'puppo'

In [29]:
df_clean['Stages'] = df_clean.apply(dog_stages, axis=1)

##### Test 

In [30]:
df_clean.head(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name  doggo  floofer  pupper  puppo Stages  
0                  10   Phineas  False    False   False  False   None  
1                  10     Tilly  False    False   False  False   None  
2                  10    Archie  False    False   False  False   None  
3                  10     Darla  False    False   False  False   None  
4                  10  Franklin  False    False   False  False   None

In [31]:
df_clean.Stages.value_counts()

pupper     224
doggo       87
puppo       24
floofer      9
Name: Stages, dtype: int64

In [32]:
df_clean = df_clean.drop(labels=['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp',
                               'pupper', 'doggo', 'puppo', 'floofer'], axis = 1)



## Issue 6: Ratings are in two columns 

#### Define: Create 1 column 'Ratings' from the denominator and nominator values 

#### Code 

In [33]:
df_clean['ratings'] = df_clean['rating_numerator']/df_clean['rating_denominator']

In [34]:
df_clean = df_clean.drop(labels=['rating_numerator', 'rating_denominator'], axis = 1)

#### Test 

In [35]:
df_clean.head(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls      name Stages  ratings  
0  https://twitter.com/dog_rates/status/892420643...   Phineas   None      1.3  
1  https://twitter.com/dog_rates/status/892177421...     Tilly   None      1.3  
2  https://twitter.com/dog_rates/status/891815181...    Archie   None      1.2  
3  https://twitter.com/dog_rates/status/891689557...     Darla   None      1.3  
4  https://twitter.com/dog_rates/status/891327558...  Franklin   None      1.2

In [36]:
df_clean.shape

(2175, 10)

### Issue #7: Some images are not dogs in the Image table

#### Define: Drop the rows that are not dog images by retaining the dog image that are True 

#### Code

In [37]:
df_imageclean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [38]:
df_imageclean = df_imageclean[df_imageclean['p1_dog'] == True]

####  Test

In [39]:
df_imageclean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1532 entries, 0 to 2073
Data columns (total 12 columns):
tweet_id    1532 non-null int64
jpg_url     1532 non-null object
img_num     1532 non-null int64
p1          1532 non-null object
p1_conf     1532 non-null float64
p1_dog      1532 non-null bool
p2          1532 non-null object
p2_conf     1532 non-null float64
p2_dog      1532 non-null bool
p3          1532 non-null object
p3_conf     1532 non-null float64
p3_dog      1532 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 124.2+ KB


In [40]:
df_imageclean.head(10)

tweet_id                                          jpg_url  \
0   666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1   666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2   666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3   666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4   666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5   666050758794694657  https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
7   666055525042405380  https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
9   666058600524156928  https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10  666063827256086533  https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11  666071193221509120  https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   

    img_num                      p1   p1_conf  p1_dog                  p2  \
0         1  Welsh_springer_spaniel  0.465074    True              collie   
1         1                 redbone  0.506826    True  miniature_pinscher   
2         1         German_shepherd  0.596461    True            malinois   
3         1     Rhodesian_ridgeback  0.408143    True             redbone   
4         1      miniature_pinscher  0.560311    True          Rottweiler   
5         1    Bernese_mountain_dog  0.651137    True    English_springer   
7         1                    chow  0.692517    True     Tibetan_mastiff   
9         1        miniature_poodle  0.201493    True            komondor   
10        1        golden_retriever  0.775930    True     Tibetan_mastiff   
11        1           Gordon_setter  0.503672    True   Yorkshire_terrier   

     p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0   0.156665    True            Shetland_sheepdog  0.061428    True  
1   0.074192    True          Rhodesian_ridgeback  0.072010    True  
2   0.138584    True                   bloodhound  0.116197    True  
3   0.360687    True           miniature_pinscher  0.222752    True  
4   0.243682    True                     Doberman  0.154629    True  
5   0.263788    True   Greater_Swiss_Mountain_dog  0.016199    True  
7   0.058279    True                     fur_coat  0.054449   False  
9   0.192305    True  soft-coated_wheaten_terrier  0.082086    True  
10  0.093718    True           Labrador_retriever  0.072427    True  
11  0.174201    True                     Pekinese  0.109454    True

## Issue 8: Timestamp is in string format 

### Define: convert timestamp data type to datetime 

#### Code 

In [41]:
df_clean["timestamp"] = pd.to_datetime(df_clean.timestamp)

#### Test 

In [42]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 10 columns):
tweet_id                 2175 non-null int64
in_reply_to_status_id    78 non-null float64
in_reply_to_user_id      78 non-null float64
timestamp                2175 non-null datetime64[ns]
source                   2175 non-null object
text                     2175 non-null object
expanded_urls            2117 non-null object
name                     2175 non-null object
Stages                   344 non-null object
ratings                  2175 non-null float64
dtypes: datetime64[ns](1), float64(3), int64(1), object(5)
memory usage: 186.9+ KB


## Tidiness Issue 

## Issue 1: Timestamp contains two variables date and time 

#### Define : split the variable into two columns "date' and 'time'

#### Code 

In [43]:
df_clean["date"] = df_clean["timestamp"].dt.date
df_clean["time"] = df_clean["timestamp"].dt.time

#### Test 

In [44]:
df_clean.head(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

            timestamp                                             source  \
0 2017-08-01 16:23:56  <a href="http://twitter.com/download/iphone" r...   
1 2017-08-01 00:17:27  <a href="http://twitter.com/download/iphone" r...   
2 2017-07-31 00:18:03  <a href="http://twitter.com/download/iphone" r...   
3 2017-07-30 15:58:51  <a href="http://twitter.com/download/iphone" r...   
4 2017-07-29 16:00:24  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls      name Stages  \
0  https://twitter.com/dog_rates/status/892420643...   Phineas   None   
1  https://twitter.com/dog_rates/status/892177421...     Tilly   None   
2  https://twitter.com/dog_rates/status/891815181...    Archie   None   
3  https://twitter.com/dog_rates/status/891689557...     Darla   None   
4  https://twitter.com/dog_rates/status/891327558...  Franklin   None   

   ratings        date      time  
0      1.3  2017-08-01  16:23:56  
1      1.3  2017-08-01  00:17:27  
2      1.2  2017-07-31  00:18:03  
3      1.3  2017-07-30  15:58:51  
4      1.2  2017-07-29  16:00:24

## Issue 2: Favorite count, retweet and image should be part of the main dataframe

### Define: merge the columns with the main dataframe  

In [45]:
#rename the tweet id in the tweet dataframe to have similar names with the rest 
df_tweetclean.rename(columns = {'id':'tweet_id'}, inplace = True)

In [46]:
df_clean2 = (pd.merge(df_clean, df_tweetclean, on='tweet_id'))

In [47]:
df_clean2.head(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   
5  891087950875897856                    NaN                  NaN   
6  890971913173991426                    NaN                  NaN   
7  890729181411237888                    NaN                  NaN   
8  890609185150312448                    NaN                  NaN   
9  890240255349198849                    NaN                  NaN   

            timestamp                                             source  \
0 2017-08-01 16:23:56  <a href="http://twitter.com/download/iphone" r...   
1 2017-08-01 00:17:27  <a href="http://twitter.com/download/iphone" r...   
2 2017-07-31 00:18:03  <a href="http://twitter.com/download/iphone" r...   
3 2017-07-30 15:58:51  <a href="http://twitter.com/download/iphone" r...   
4 2017-07-29 16:00:24  <a href="http://twitter.com/download/iphone" r...   
5 2017-07-29 00:08:17  <a href="http://twitter.com/download/iphone" r...   
6 2017-07-28 16:27:12  <a href="http://twitter.com/download/iphone" r...   
7 2017-07-28 00:22:40  <a href="http://twitter.com/download/iphone" r...   
8 2017-07-27 16:25:51  <a href="http://twitter.com/download/iphone" r...   
9 2017-07-26 15:59:51  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   
5  Here we have a majestic great white breaching ...   
6  Meet Jax. He enjoys ice cream so much he gets ...   
7  When you watch your owner call another dog a g...   
8  This is Zoey. She doesn't want to be one of th...   
9  This is Cassie. She is a college pup. Studying...   

                                       expanded_urls      name Stages  \
0  https://twitter.com/dog_rates/status/892420643...   Phineas   None   
1  https://twitter.com/dog_rates/status/892177421...     Tilly   None   
2  https://twitter.com/dog_rates/status/891815181...    Archie   None   
3  https://twitter.com/dog_rates/status/891689557...     Darla   None   
4  https://twitter.com/dog_rates/status/891327558...  Franklin   None   
5  https://twitter.com/dog_rates/status/891087950...       Dog   None   
6  https://gofundme.com/ydvmve-surgery-for-jax,ht...       Jax   None   
7  https://twitter.com/dog_rates/status/890729181...       Dog   None   
8  https://twitter.com/dog_rates/status/890609185...      Zoey   None   
9  https://twitter.com/dog_rates/status/890240255...    Cassie  doggo   

   ratings        date      time  retweet_count  favorite_count  
0      1.3  2017-08-01  16:23:56           8853           39467  
1      1.3  2017-08-01  00:17:27           6514           33819  
2      1.2  2017-07-31  00:18:03           4328           25461  
3      1.3  2017-07-30  15:58:51           8964           42908  
4      1.2  2017-07-29  16:00:24           9774           41048  
5      1.3  2017-07-29  00:08:17           3261           20562  
6      1.3  2017-07-28  16:27:12           2158           12041  
7      1.3  2017-07-28  00:22:40          16716           56848  
8      1.3  2017-07-27  16:25:51           4429           28226  
9      1.4  2017-07-26  15:59:51           7711           32467

In [48]:
df_clean2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2174
Data columns (total 14 columns):
tweet_id                 2175 non-null int64
in_reply_to_status_id    78 non-null float64
in_reply_to_user_id      78 non-null float64
timestamp                2175 non-null datetime64[ns]
source                   2175 non-null object
text                     2175 non-null object
expanded_urls            2117 non-null object
name                     2175 non-null object
Stages                   344 non-null object
ratings                  2175 non-null float64
date                     2175 non-null object
time                     2175 non-null object
retweet_count            2175 non-null int64
favorite_count           2175 non-null int64
dtypes: datetime64[ns](1), float64(3), int64(3), object(7)
memory usage: 254.9+ KB


In [49]:
#drop columns that are not needed in the image dataframe 
df_imageclean2 = df_imageclean.drop(labels=['p1', 'p1_conf', 'p1_dog', 'p3', 'p3_dog', 'p3_conf', 
                               'p2', 'p2_conf', 'p2_dog'], axis = 1)

In [50]:
df_clean3 = (pd.merge(df_clean2, df_imageclean2, on='tweet_id'))

In [51]:
df_clean3.head(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892177421306343426                    NaN                  NaN   
1  891815181378084864                    NaN                  NaN   
2  891327558926688256                    NaN                  NaN   
3  891087950875897856                    NaN                  NaN   
4  890971913173991426                    NaN                  NaN   

            timestamp                                             source  \
0 2017-08-01 00:17:27  <a href="http://twitter.com/download/iphone" r...   
1 2017-07-31 00:18:03  <a href="http://twitter.com/download/iphone" r...   
2 2017-07-29 16:00:24  <a href="http://twitter.com/download/iphone" r...   
3 2017-07-29 00:08:17  <a href="http://twitter.com/download/iphone" r...   
4 2017-07-28 16:27:12  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Tilly. She's just checking pup on you....   
1  This is Archie. He is a rare Norwegian Pouncin...   
2  This is Franklin. He would like you to stop ca...   
3  Here we have a majestic great white breaching ...   
4  Meet Jax. He enjoys ice cream so much he gets ...   

                                       expanded_urls      name Stages  \
0  https://twitter.com/dog_rates/status/892177421...     Tilly   None   
1  https://twitter.com/dog_rates/status/891815181...    Archie   None   
2  https://twitter.com/dog_rates/status/891327558...  Franklin   None   
3  https://twitter.com/dog_rates/status/891087950...       Dog   None   
4  https://gofundme.com/ydvmve-surgery-for-jax,ht...       Jax   None   

   ratings        date      time  retweet_count  favorite_count  \
0      1.3  2017-08-01  00:17:27           6514           33819   
1      1.2  2017-07-31  00:18:03           4328           25461   
2      1.2  2017-07-29  16:00:24           9774           41048   
3      1.3  2017-07-29  00:08:17           3261           20562   
4      1.3  2017-07-28  16:27:12           2158           12041   

                                           jpg_url  img_num  
0  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg        1  
1  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg        1  
2  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg        2  
3  https://pbs.twimg.com/media/DF3HwyEWsAABqE6.jpg        1  
4  https://pbs.twimg.com/media/DF1eOmZXUAALUcq.jpg        1

In [52]:
df_clean3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1477 entries, 0 to 1476
Data columns (total 16 columns):
tweet_id                 1477 non-null int64
in_reply_to_status_id    14 non-null float64
in_reply_to_user_id      14 non-null float64
timestamp                1477 non-null datetime64[ns]
source                   1477 non-null object
text                     1477 non-null object
expanded_urls            1477 non-null object
name                     1477 non-null object
Stages                   228 non-null object
ratings                  1477 non-null float64
date                     1477 non-null object
time                     1477 non-null object
retweet_count            1477 non-null int64
favorite_count           1477 non-null int64
jpg_url                  1477 non-null object
img_num                  1477 non-null int64
dtypes: datetime64[ns](1), float64(3), int64(4), object(8)
memory usage: 196.2+ KB


In [53]:
#drop columns not needed
df_clean3 = df_clean3.drop(labels=['in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp'], axis = 1)

In [54]:
df_clean3

tweet_id                                             source  \
0     892177421306343426  <a href="http://twitter.com/download/iphone" r...   
1     891815181378084864  <a href="http://twitter.com/download/iphone" r...   
2     891327558926688256  <a href="http://twitter.com/download/iphone" r...   
3     891087950875897856  <a href="http://twitter.com/download/iphone" r...   
4     890971913173991426  <a href="http://twitter.com/download/iphone" r...   
5     890729181411237888  <a href="http://twitter.com/download/iphone" r...   
6     890609185150312448  <a href="http://twitter.com/download/iphone" r...   
7     890240255349198849  <a href="http://twitter.com/download/iphone" r...   
8     890006608113172480  <a href="http://twitter.com/download/iphone" r...   
9     889880896479866881  <a href="http://twitter.com/download/iphone" r...   
10    889665388333682689  <a href="http://twitter.com/download/iphone" r...   
11    889638837579907072  <a href="http://twitter.com/download/iphone" r...   
12    889531135344209921  <a href="http://twitter.com/download/iphone" r...   
13    889278841981685760  <a href="http://twitter.com/download/iphone" r...   
14    888917238123831296  <a href="http://twitter.com/download/iphone" r...   
15    888804989199671297  <a href="http://twitter.com/download/iphone" r...   
16    888554962724278272  <a href="http://twitter.com/download/iphone" r...   
17    888078434458587136  <a href="http://twitter.com/download/iphone" r...   
18    887705289381826560  <a href="http://twitter.com/download/iphone" r...   
19    887473957103951883  <a href="http://twitter.com/download/iphone" r...   
20    887343217045368832  <a href="http://twitter.com/download/iphone" r...   
21    887101392804085760  <a href="http://twitter.com/download/iphone" r...   
22    886983233522544640  <a href="http://twitter.com/download/iphone" r...   
23    886736880519319552  <a href="http://twitter.com/download/iphone" r...   
24    886366144734445568  <a href="http://twitter.com/download/iphone" r...   
25    886258384151887873  <a href="http://twitter.com/download/iphone" r...   
26    885984800019947520  <a href="http://twitter.com/download/iphone" r...   
27    885528943205470208  <a href="http://twitter.com/download/iphone" r...   
28    885167619883638784  <a href="http://twitter.com/download/iphone" r...   
29    884925521741709313  <a href="http://twitter.com/download/iphone" r...   
...                  ...                                                ...   
1447  666644823164719104  <a href="http://twitter.com/download/iphone" r...   
1448  666454714377183233  <a href="http://twitter.com/download/iphone" r...   
1449  666447344410484738  <a href="http://twitter.com/download/iphone" r...   
1450  666437273139982337  <a href="http://twitter.com/download/iphone" r...   
1451  666435652385423360  <a href="http://twitter.com/download/iphone" r...   
1452  666428276349472768  <a href="http://twitter.com/download/iphone" r...   
1453  666421158376562688  <a href="http://twitter.com/download/iphone" r...   
1454  666418789513326592  <a href="http://twitter.com/download/iphone" r...   
1455  666407126856765440  <a href="http://twitter.com/download/iphone" r...   
1456  666396247373291520  <a href="http://twitter.com/download/iphone" r...   
1457  666373753744588802  <a href="http://twitter.com/download/iphone" r...   
1458  666353288456101888  <a href="http://twitter.com/download/iphone" r...   
1459  666345417576210432  <a href="http://twitter.com/download/iphone" r...   
1460  666287406224695296  <a href="http://twitter.com/download/iphone" r...   
1461  666273097616637952  <a href="http://twitter.com/download/iphone" r...   
1462  666102155909144576  <a href="http://twitter.com/download/iphone" r...   
1463  666099513787052032  <a href="http://twitter.com/download/iphone" r...   
1464  666094000022159362  <a href="http://twitter.com/download/iphone" r...   
1465  666082916733198337  <a href="http://twitter.com/download/iph

## Storing Data
Save gathered, assessed, and cleaned master dataset to a CSV file named "twitter_archive_master.csv".

In [55]:
df_clean3.to_csv('twitter_archive_master.csv', index=False)

## Analyzing and Visualizing Data
In this section, analyze and visualize your wrangled data. You must produce at least **three (3) insights and one (1) visualization.**

In [56]:
df = pd.read_csv('twitter_archive_master.csv')

In [57]:
df.head(5)

tweet_id                                             source  \
0  892177421306343426  <a href="http://twitter.com/download/iphone" r...   
1  891815181378084864  <a href="http://twitter.com/download/iphone" r...   
2  891327558926688256  <a href="http://twitter.com/download/iphone" r...   
3  891087950875897856  <a href="http://twitter.com/download/iphone" r...   
4  890971913173991426  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Tilly. She's just checking pup on you....   
1  This is Archie. He is a rare Norwegian Pouncin...   
2  This is Franklin. He would like you to stop ca...   
3  Here we have a majestic great white breaching ...   
4  Meet Jax. He enjoys ice cream so much he gets ...   

                                       expanded_urls      name Stages  \
0  https://twitter.com/dog_rates/status/892177421...     Tilly    NaN   
1  https://twitter.com/dog_rates/status/891815181...    Archie    NaN   
2  https://twitter.com/dog_rates/status/891327558...  Franklin    NaN   
3  https://twitter.com/dog_rates/status/891087950...       Dog    NaN   
4  https://gofundme.com/ydvmve-surgery-for-jax,ht...       Jax    NaN   

   ratings        date      time  retweet_count  favorite_count  \
0      1.3  2017-08-01  00:17:27           6514           33819   
1      1.2  2017-07-31  00:18:03           4328           25461   
2      1.2  2017-07-29  16:00:24           9774           41048   
3      1.3  2017-07-29  00:08:17           3261           20562   
4      1.3  2017-07-28  16:27:12           2158           12041   

                                           jpg_url  img_num  
0  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg        1  
1  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg        1  
2  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg        2  
3  https://pbs.twimg.com/media/DF3HwyEWsAABqE6.jpg        1  
4  https://pbs.twimg.com/media/DF1eOmZXUAALUcq.jpg        1

In [58]:
df.describe()

tweet_id      ratings  retweet_count  favorite_count      img_num
count  1.477000e+03  1477.000000    1477.000000     1477.000000  1477.000000
mean   7.406665e+17     1.156534    2790.399458     9288.481381     1.213947
std    6.860818e+16     0.788651    4763.557005    12593.129577     0.575230
min    6.660209e+17     0.100000      16.000000       81.000000     1.000000
25%    6.782555e+17     1.000000     656.000000     2232.000000     1.000000
50%    7.152202e+17     1.100000    1443.000000     4503.000000     1.000000
75%    7.954003e+17     1.200000    3261.000000    11746.000000     1.000000
max    8.921774e+17    16.500000   79515.000000   132810.000000     4.000000

In [59]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1477 entries, 0 to 1476
Data columns (total 13 columns):
tweet_id          1477 non-null int64
source            1477 non-null object
text              1477 non-null object
expanded_urls     1477 non-null object
name              1477 non-null object
Stages            228 non-null object
ratings           1477 non-null float64
date              1477 non-null object
time              1477 non-null object
retweet_count     1477 non-null int64
favorite_count    1477 non-null int64
jpg_url           1477 non-null object
img_num           1477 non-null int64
dtypes: float64(1), int64(4), object(8)
memory usage: 150.1+ KB


In [60]:
df.Stages.value_counts()

pupper     144
doggo       58
puppo       19
floofer      7
Name: Stages, dtype: int64

### Insights:
1. there are 1477 rows and 13 columns of weratedog data 

2. All the dogs has at least one image with some having four images 

3. The maximum rating a dog has had is  16.5

### Visualization

In [61]:
df['Stages'].value_counts().plot(kind='bar')